### PR TITLE
[mds-provider] event processing

### PR DESCRIPTION
**Pull request by mdurling**
Latest enhancements and bug fixes.

## Impacts
- [x] Provider
- [ ] Agency
- [ ] Audit
- [ ] Policy
- [ ] Compliance

### CODE REVIEW

File: packages/mds-db/index.ts [hunk 1]
```diff
@@ -1513,16 +1522,19 @@ async function readEventsRangeExclusive(
...
   )
   return rows.map(({ lat, lng, telemetry_timestamp, ...event }) => ({
     ...event,
     telemetry_timestamp,
-    telemetry: {
-      timestamp: telemetry_timestamp,
-      gps: { lat, lng }
-    }
+    telemetry:
+      lat !== null && lng !== null
```
Comment (avatarneil) [inline packages/mds-db/index.ts]: Could just be: `telemetry: lat && lng ? { timestamp: telemetry_timestamp, gps: { lat, lng } } : null`

File: packages/mds-provider/api.ts [hunk 2]
```diff
@@ -221,66 +204,56 @@ function api(app: express.Express): express.Express {
...
-      properties: {
-        timestamp: item.timestamp
-      },
-      geometry: {
-        type: 'Point',
-        coordinates: [round(item.gps.lng, 6), round(item.gps.lat, 6)]
-      }
-    }
-  }
+  // function asFeature(item: Telemetry): Feature {
```
Comment (avatarneil) [inline packages/mds-provider/api.ts]: Why commenting out versus nuking/moving to `mds-utils`?

File: packages/mds-provider/api.ts [hunk 3]
```diff
@@ -221,66 +204,56 @@ function api(app: express.Express): express.Express {
     return undefined
```
Comment (avatarneil) [inline packages/mds-provider/api.ts]: We don't need to return anything here I think

File: packages/mds-provider/event-processor/index.ts [hunk 4]
```diff
@@ -123,28 +117,53 @@ async function process(options: ReadStreamOptions): Promise<void> {
...
+const stop = async (): Promise<void> => {
   logger.info('Stopping Event Processor')
   await Promise.all([stream.shutdown(), db.shutdown()])
 }
 
-export type ProviderEventProcessorOptions = ReadStreamOptions
+const wait = async (interval: number): Promise<number> =>
+  interval > 0 ? new Promise(resolve => setTimeout(() => resolve(interval), interval)) : Promise.resolve(interval)
+
+/* eslint-disable no-await-in-loop */
```
Comment (avatarneil) [inline packages/mds-provider/event-processor/index.ts]: Add an `/* eslint-reason ...*/` please

File: packages/mds-provider/package.json [hunk 5]
```diff
@@ -14,8 +14,8 @@
     "test": "yarn test:prettier && yarn test:eslint && yarn test:unit",
     "test:eslint": "eslint --ignore-path ../../.gitignore '**/*.ts'",
     "test:prettier": "prettier --write --check --ignore-path ../../.gitignore '**/*.ts'",
-    "test:unit": "nyc --check-coverage --exclude tests --extension .ts --lines 80 --reporter=text --reporter=html ts-mocha --require source-map-support/register --recursive tests/**/*.ts",
-    "watch": "nodemon --watch '../../packages' --ext 'ts' --ignore '*.d.ts' --exec yarn watch:exec",
+    "test:unit": "nyc --check-coverage --exclude tests --extension .ts --lines 75 --reporter=text --reporter=html ts-mocha --require source-map-support/register --recursive tests/**/*.ts",
```
Comment (avatarneil) [inline packages/mds-provider/package.json]: I don't really like the idea of reducing test coverage % significantly with the addition of new features... We should either write new tests which have more coverage, or add istanbul annotations if the code can't be reasonably tested.

File: packages/mds-db/index.ts (hunk 1 above)
Comment (mdurling) [reply packages/mds-db/index.ts]: I guess that's safe ... as long as you want to treat `{ lat: 0, lng: 0 }` as `null` also.

File: packages/mds-provider/api.ts (hunk 2 above)
Comment (mdurling) [reply packages/mds-provider/api.ts]: They will be making a comeback.

File: packages/mds-provider/api.ts (hunk 3 above)
Comment (mdurling) [reply packages/mds-provider/api.ts]: Actually, I think `return {}` is probably more appropriate.

File: packages/mds-provider/package.json (hunk 5 above)
Comment (mdurling) [reply packages/mds-provider/package.json]: I did add some new tests and increased the % ... I will update this to `--lines 85`

File: packages/mds-db/index.ts (hunk 1 above)
Comment (karcass) [reply packages/mds-db/index.ts]: or undefined :)

Comment (mdurling) [reply packages/mds-db/index.ts]: Coming from db it will never be undefined

Comment (karcass) [reply packages/mds-db/index.ts]: doh

File: packages/mds-db/index.ts [hunk 6]
```diff
@@ -1266,6 +1266,14 @@ async function readStatusChanges(
     }
   }
 
+  if (device_id) {
+    if (!isUUID(device_id)) {
```
Comment (janedotx) [inline packages/mds-db/index.ts]: Real dumb question, but isn't the type declaration supposed to ensure this is a UUID?

Comment (mdurling) [reply packages/mds-db/index.ts]: If I understand your question correctly, ... The `UUID` type is just an alias for `string`. Regardless though, static type checking only helps at compilation time. At run time, you have to type guard everything. That's why most of the validators I've written take an input argument of type `arg: unknown` and returns `arg is <type>`. Which is basically, accept this argument of unknown type, apply some tests/constraints to it, and it they all pass, treat it as a `<type>` from that point forward in the code.

Comment (janedotx) [reply packages/mds-db/index.ts]: Ah right, that makes sense. Thanks.

**Pull request merged**